cpp/src/arrow/compute/kernels/scalar_fill_null_test.cc

```diff
@@ -0,0 +1,137 @@
+  auto datum = Datum(std::make_shared<NullScalar>());
+  CheckFillNull<NullType, std::nullptr_t>(null(), {0, 0, 0, 0},
+                                          {false, false, false, false}, datum,
+                                          {0, 0, 0, 0}, {false, false, false, false});
+  CheckFillNull<NullType, std::nullptr_t>(null(), {NULL, NULL, NULL, NULL}, {}, datum,
+                                          {NULL, NULL, NULL, NULL}, {});
+  CheckFillNull<NullType, std::nullptr_t>(null(), {0, 0, 0, 0},
+                                          {false, false, false, false}, datum,
+                                          {0, 0, 0, 0}, {false, false, false, false});
+  CheckFillNull<NullType, std::nullptr_t>(null(), {NULL, NULL, NULL, NULL}, {}, datum,
+                                          {NULL, NULL, NULL, NULL}, {});
+}
+
+TEST_F(TestFillNullKernel, FillNullBoolean) {
+  auto scalar1 = std::make_shared<BooleanScalar>(false);
+  auto scalar2 = std::make_shared<BooleanScalar>(true);
+  // no nulls
+  CheckFillNull<BooleanType, bool>(boolean(), {true, false, true, false},
+                                   {true, true, true, true}, Datum(scalar1),
+                                   {true, false, true, false}, {true, true, true, true});
+  // some nulls
+  CheckFillNull<BooleanType, bool>(boolean(), {true, false, true, false},
+                                   {false, true, true, false}, Datum(scalar1),
+                                   {false, false, true, false}, {true, true, true, true});
+  CheckFillNull<BooleanType, bool>(boolean(), {true, false, true, false},
+                                   {false, true, false, false}, Datum(scalar2),
+                                   {true, false, true, true}, {true, true, true, true});
+}
+
+TEST_F(TestFillNullKernel, FillNullTimeStamp) {
+  auto time32_type = time32(TimeUnit::SECOND);
+  auto time64_type = time64(TimeUnit::NANO);
+  auto scalar1 = Datum(std::make_shared<Time32Scalar>(5, time32_type));
+  auto scalar2 = Datum(std::make_shared<Time64Scalar>(6, time64_type));
+  // no nulls
+  CheckFillNull<Time32Type, int32_t>(time32_type, {2, 1, 6, 9}, {true, true, true, true},
+                                     Datum(scalar1), {2, 1, 6, 9},
+                                     {true, true, true, true});
+  CheckFillNull<Time32Type, int32_t>(time32_type, {2, 1, 6, 9},
+                                     {true, false, true, false}, Datum(scalar1),
+                                     {2, 5, 6, 5}, {true, true, true, true});
+  // some nulls
+  CheckFillNull<Time64Type, int64_t>(time64_type, {2, 1, 6, 9}, {true, true, true, true},
+                                     scalar2, {2, 1, 6, 9}, {true, true, true, true});
+  CheckFillNull<Time64Type, int64_t>(time64_type, {2, 1, 6, 9},
+                                     {true, false, true, false}, scalar2, {2, 6, 6, 6},
+                                     {true, true, true, true});
+}
+
```

Comment:
The behavior of the kernel when passing a scalar with `is_valid=false` is not validated (and probably yield incorrect results)

Reply:
I have included a test case